Assets/HoloToolkit/Input/Scripts/Interactions/TapToPlace.cs

```diff
@@ -123,86 +152,99 @@ private void HandlePlacement()
-                    SpatialMappingManager.Instance.DrawVisualMeshes = true;
-                }
+        private void StartPlacing()
+        {
+            var layerCacheTarget = PlaceParentOnTap ? ParentGameObjectToPlace : gameObject;
+            layerCacheTarget.SetLayerRecursively(IgnoreRaycastLayer, out layerCache);
+            InputManager.Instance.PushModalInputHandler(gameObject);
+
+            ToggleSpatialMesh();
 #if UNITY_WSA && !UNITY_EDITOR
 
-                //Removes existing world anchor if any exist.
-                WorldAnchorManager.Instance.RemoveAnchor(gameObject);
+            //Removes existing world anchor if any exist.
+            WorldAnchorManager.Instance.RemoveAnchor(gameObject);
 #endif
-            }
-            else
-            {
-                SetLayerRecursively(transform, useDefaultLayer: true);
-                // Clear our cache in case we added or removed gameobjects between taps
-                defaultLayersCache.Clear();
-                InputManager.Instance.RemoveGlobalListener(gameObject);
+        }
 
-                // If the user is not in placing mode, hide the spatial mapping mesh.
-                if (AllowMeshVisualizationControl)
-                {
-                    SpatialMappingManager.Instance.DrawVisualMeshes = false;
-                }
+        private void StopPlacing()
+        {
+            var layerCacheTarget = PlaceParentOnTap ? ParentGameObjectToPlace : gameObject;
+            layerCacheTarget.ApplyLayerCacheRecursively(layerCache);
+            InputManager.Instance.PopModalInputHandler();
+
+            ToggleSpatialMesh();
 #if UNITY_WSA && !UNITY_EDITOR
 
-                // Add world anchor when object placement is done.
-                WorldAnchorManager.Instance.AttachAnchor(gameObject, SavedAnchorFriendlyName);
+            // Add world anchor when object placement is done.
+            WorldAnchorManager.Instance.AttachAnchor(gameObject, SavedAnchorFriendlyName);
 #endif
-            }
         }
 
-        private void DetermineParent()
+        /// <summary>
+        /// If the user is in placing mode, display the spatial mapping mesh.
+        /// </summary>
+        private void ToggleSpatialMesh()
         {
-            if (!PlaceParentOnTap) { return; }
-
-            if (ParentGameObjectToPlace == null)
-            {
-                if (gameObject.transform.parent == null)
-                {
-                    Debug.LogWarning("The selected GameObject has no parent.");
-                    PlaceParentOnTap = false;
-                }
-                else
-                {
-                    Debug.LogWarning("No parent specified. Using immediate parent instead: " + gameObject.transform.parent.gameObject.name);
-                    ParentGameObjectToPlace = gameObject.transform.parent.gameObject;
-                }
-            }
+            SpatialMappingManager.Instance.DrawVisualMeshes = IsBeingPlaced && AllowMeshVisualizationControl;
```

Comment:
`SpatialMappingManager` is also optional. We should check if we have a valid instance.